### PR TITLE
PML-286 g2 noise

### PR DESCRIPTION
**Pull request by LF-Vigneux**
## Summary
Added the g2 noise simulations in the MerLin's quantum layer's pipeline. The simulations can now be .

## Related Issue
PML-286

## Type of change
- [ ] Bug fix
- [x] New feature
- [x] Documentation update
- [ ] Refactor / Cleanup
- [ ] Performance improvement
- [ ] CI / Build / Tooling
- [ ] Breaking change (requires migration notes)

## Proposed changes
- There is now a NoisyG2SLOSComputeGraph class that is used whenthere is g2 noise.
- This class simulates the probs of the circuit with the g2 noise.
- The same modifications in the pipeline as the NoisySLOSComputeGraph for indistinguishability wew done plus:
     - Added a guard preventing grouping strategies with g2 noises as the output is not a tensor and is sector dependant.
     - Changed the sampling function handling in the quantum layer. We do not pass a sampling function anymore but a class that has a normal sampler and a g2 sampler that samples across all photon sectors.  
- Added new types to handle those photon sectors.

## How to test / How to run

1. Command lines

```
pytest -q
```

## Performance considerations (optional)
- G2 SLOS simulations are heavier as expected

## Documentation
- [x] User docs updated (Sphinx)
- [ ] Examples / notebooks updated
- [x] Docstrings updated
- [x] Updated the API

### CODE REVIEW

**Comment by CassNot** [comment]
@codex review with a focus on scaling and memory usage

File: merlin/pcvl_pytorch/slos_torchscript.py [hunk 1]
```diff
@@ -1151,15 +1154,20 @@ def build_slos_distribution_computegraph(
...
-        compute_graph = NoisySLOSComputeGraph(
-            noise_groups,
-            m,
-            n_photons,
-            computation_space,
-            keep_keys,
-            device,
-            dtype,
-        )
+        if "g2" in noise_groups.source:
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/pcvl_pytorch/slos_torchscript.py:1157]: This should not happen anyways as it is checked by the classify noise model method

File: merlin/measurement/process.py [hunk 2]
```diff
@@ -133,6 +134,111 @@ def pcvl_sampler(
...
+        for sector in distribution.sectors:
+            if one_dimension_vector is None:
+                one_dimension_vector = sector.tensor
+                if one_dimension_vector.dim() == 1:
+                    augment_input = True
+                    one_dimension_vector.unsqueeze(0)
+                else:
+                    augment_input = False
+                indexes.append((0, one_dimension_vector.size(1)))
+            else:
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/measurement/process.py:181]: Fixed in next commit

File: merlin/pcvl_pytorch/noisy_slos.py [hunk 3]
```diff
@@ -24,13 +26,328 @@
...
+        num_input_photons = sum(input_state)
+        # Convert g2 = g^(2)(0) second-order coherence to per-source emission
+        # probability p.  The two are related by g^(2)(0) = 2p/(1+p)^2, so:
+        #   p = ((1 - g2) - sqrt(1 - 2*g2)) / g2,  valid for g2 in [0, 0.5].
+        # For small g2, p ≈ g2/2 (L'Hôpital).  At g2=0 the limit is p=0.
+        _g2 = (
+            self.g2
+            if isinstance(self.g2, torch.Tensor)
+            else torch.tensor(float(self.g2))
+        )
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/pcvl_pytorch/noisy_slos.py]: Changed in next commit

File: merlin/pcvl_pytorch/noisy_slos.py [hunk 4]
```diff
@@ -24,13 +26,328 @@
...
+            # Creating the n_photons + num_photons_added sector
+            sector = SectorResult(
+                torch.zeros(
+                    Combinadics(
+                        scheme="fock", n=self.n_photons + num_photons_added, m=self.m
+                    ).compute_space_size()
+                ),
+                n_modes=self.m,
+                n_photons=self.n_photons + num_photons_added,
+            )
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/pcvl_pytorch/noisy_slos.py:226]: Fixed in next commit

File: merlin/measurement/strategies.py [hunk 5]
```diff
@@ -120,24 +122,38 @@ def supports_sampling(self) -> bool:
...
         grouping: Callable[[torch.Tensor], torch.Tensor] | None = None,
-    ) -> torch.Tensor:
+    ) -> torch.Tensor | SectoredDistribution:
         # Distribution strategies apply detector/noise transforms before sampling.
-        distribution = apply_photon_loss(distribution)
-        distribution = apply_detectors(distribution)
+        if isinstance(distribution, SectoredDistribution):
+            if grouping:
+                raise RuntimeError(
+                    f"A grouping strategy can not be applied to the output of a simulation with g2 noise. Indeed, since this noise creates input states with more phtons than expected, multiple photon sectors are explored. The fock spaces explored are m={distribution.sectors[0].n_modes} modes and n_photons={min(distribution._photon_map.keys())} to 2*n_photons={max(distribution._photon_map.keys())} that all have different space dimensions. To still apply a grouping strategy, you can iterate over the :class:`~merlin.core.sectored_distribution.SectorResult`s of the :class:`~merlin.core.sectored_distribution.SectoredDistribution` and apply one grouping per sector."
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/measurement/strategies.py]: Changed in next commit

File: docs/source/quantum_expert_area/noisy_simulations.rst [hunk 6]
```diff
@@ -88,6 +88,7 @@ For noisy simulations, there are a couple of rules that need to be followed:
 1. All noisy simulations must be run with the probabilities measurement strategy.
 2. Noisy simulations cannot use ``return_object=True``.
 3. Noisy simulations with source noise must be run in the Fock computation space. If a different space is chosen, it will be changed automatically with a warning.
+4. Noisy simulation with ``g2>0`` can not use a grouping strategy. Indeed, since this noise creates input states with more phtons than expected, multiple photon sectors are explored. The fock spaces explored are m modes and n_photons to 2*n_photons that all have different space dimensions. To still apply a grouping strategy, you can iterate over the :class:`~merlin.core.sectored_distribution.SectorResult` objects of the :class:`~merlin.core.sectored_distribution.SectoredDistribution` and apply one grouping per sector.
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply docs/source/quantum_expert_area/noisy_simulations.rst:92]: Changed in next commit

File: tests/algorithms/test_noise_contract.py [hunk 7]
```diff
@@ -216,64 +216,33 @@ def test_noisy_layer_with_probs_strategy_raises_not_implemented():
...
-                    g2=0.2,
-                ),
-                measurement_strategy=ml.MeasurementStrategy.probs(
-                    computation_space=ml.ComputationSpace.FOCK
-                ),
-            )
+            measurement_strategy=ml.MeasurementStrategy.probs(
+                computation_space=ml.ComputationSpace.FOCK
+            ),
+        )
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply tests/algorithms/test_noise_contract.py]: Simplified the test for only non implemented noise/ errors

File: merlin/core/sectored_distribution.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+    def get_sector(self, n_photons: int) -> SectorResult:
+        """Return the SectorResult associated with n_photons."""
+        if n_photons not in self._photon_map.keys():
+            raise ValueError("No SectorResult with that number of photons")
+        return self.sectors[self._photon_map[n_photons]]
+
+    def total_probability(self) -> float:
+        """Return the SectorResult associated with n_photons."""
+        total_prob = 0.0
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/core/sectored_distribution.py]: Changed in next commit

File: merlin/measurement/strategies.py [hunk 9]
```diff
@@ -74,28 +75,29 @@ def process(
...
+        amplitudes: torch.Tensor | SectoredDistribution,
         apply_sampling: bool,
         effective_shots: int,
-        sample_fn: Callable[[torch.Tensor, int], torch.Tensor],
+        sampler: SamplingProcess,
         apply_photon_loss: Callable[[torch.Tensor], torch.Tensor],
         apply_detectors: Callable[[torch.Tensor], torch.Tensor],
         grouping: Callable[[torch.Tensor], torch.Tensor] | None = None,
-    ) -> torch.Tensor | PartialMeasurement:
+    ) -> torch.Tensor | PartialMeasurement | SectoredDistribution:
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/measurement/strategies.py:89]: Fixed in next commit

File: merlin/pcvl_pytorch/noisy_slos.py [hunk 10]
```diff
@@ -24,13 +26,328 @@
...
+                self.computation_space.casefold(), n=self.n_photons, m=self.m
+            ).enumerate_states()
+        ]
+
+    def _get_extra_photon_combinations(
+        self,
+        input_state: list[int] | tuple[int, ...],
+    ) -> list[list[tuple[int]]]:
+        num_photons = sum(input_state)
+        output: list[list[tuple[int]]] = [[]]
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/pcvl_pytorch/noisy_slos.py]: Fixed in next commit

File: merlin/core/process.py [hunk 11]
```diff
@@ -208,36 +216,96 @@ def compute(
...
                 active_indices = torch.nonzero(active_mask, as_tuple=True)[0].tolist()
-
-                probs_per_state = []
-                for idx in active_indices:
-                    input_fock_state = self.simulation_graph.mapped_keys[idx]
-                    _, probs = self.simulation_graph.compute_probs(
-                        unitary, input_fock_state
+                # If there is g2 noise, the data needs to be treated per photon sector
+                if isinstance(self.simulation_graph, NoisyG2SLOSComputeGraph):
+                    probs_per_state = []
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply merlin/core/process.py]: Changed in next commit

File: merlin/measurement/strategies.py [hunk 12]
```diff
@@ -120,24 +123,38 @@ def supports_sampling(self) -> bool:
...
         # Distribution strategies apply detector/noise transforms before sampling.
-        distribution = apply_photon_loss(distribution)
-        distribution = apply_detectors(distribution)
+        if isinstance(distribution, SectoredDistribution):
+            if grouping:
+                raise RuntimeError(
+                    f"A grouping strategy cannot be applied to the output of a simulation with g2 noise. Indeed, since this noise creates input states with more photons than expected, multiple photon sectors are explored. The fock spaces explored are m={distribution.sectors[0].n_modes} modes and n_photons={min(distribution._photon_map.keys())} to 2*n_photons={max(distribution._photon_map.keys())} that all have different space dimensions. To still apply a grouping strategy, you can iterate over the :class:`~merlin.core.sectored_distribution.SectorResult`s of the :class:`~merlin.core.sectored_distribution.SectoredDistribution` and apply one grouping per sector."
+                )
+            for sector_result in distribution.sectors:
+                sector_result.tensor = apply_photon_loss(sector_result.tensor)
```
Comment (ben9871) [inline merlin/measurement/strategies.py]: This reuses the photon-loss and detector transforms built from the original `n_photons` Fock basis for all g2 sectors. Higher sectors have different basis-sizes, so post-measurement loss/detectors either crash or would map against the wrong basis.

File: merlin/pcvl_pytorch/noisy_slos.py [hunk 13]
```diff
@@ -24,13 +26,326 @@
...
+        _g2 = torch.as_tensor(self.g2, device=self.device, dtype=self.dtype)
+        _disc = (1.0 - 2.0 * _g2).clamp(min=0.0)
+        p_emit = ((1.0 - _g2) - _disc.sqrt()) / _g2.clamp(min=1e-15)
+        for num_photons_added in range(len(extra_photons_combinations)):
+            weight_k = (p_emit**num_photons_added) * (
+                (1 - p_emit) ** (num_input_photons - num_photons_added)
+            )
+
+            # Creating the n_photons + num_photons_added sector
+            sector = SectorResult(
```
Comment (ben9871) [inline merlin/pcvl_pytorch/noisy_slos.py:216]: `SectorResult.keys` is never populated, so the output does not actually carry the sector basis metadata promised by the API. This is also why the measurement layer cannot safely build per-sector loss/detector transforms.

File: merlin/pcvl_pytorch/noisy_slos.py [hunk 14]
```diff
@@ -24,13 +26,326 @@
...
+        # Convert to tensor if not already
+        if not isinstance(input_state, Tensor):
+            input_state_tensor = torch.tensor(list(input_state), dtype=torch.int32)
+        else:
+            input_state_tensor = input_state.int()
+
+        positions = torch.arange(len(input_state_tensor), dtype=torch.long).to(
+            self.device
+        )
+        photon_positions = torch.repeat_interleave(positions, input_state_tensor).to(
```
Comment (ben9871) [inline merlin/pcvl_pytorch/noisy_slos.py:156]: This enumerates g2 events per photon, not per active source mode. For a bunched input like `[2, 0, 0]`, the graph emits sectors `[2, 3, 4]` even though only mode 0 can fire a g2 event.

File: merlin/core/sectored_distribution.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        """Return the SectorResult associated with n_photons."""
+        if n_photons not in self._photon_map.keys():
+            raise ValueError("No SectorResult with that number of photons")
+        return self.sectors[self._photon_map[n_photons]]
+
+    def total_probability(self) -> float:
+        """Returns the total probability across the sectors."""
+        total_prob = 0.0
+        for sector in self.sectors:
+            total_prob += sector.tensor.sum().item()
```
Comment (ben9871) [inline merlin/core/sectored_distribution.py]: `total_probability()` should return a tensor, not a Python float. The `.item()` calls detach the result from autograd, which works against the differentiable output contract.

**Review by ben9871 — CHANGES_REQUESTED** [review]
Some issues with behaviour of G2 and accessibility of keys to be fixed before merging

File: merlin/pcvl_pytorch/noisy_slos.py (hunk 14 above)
Comment (LF-Vigneux) [reply merlin/pcvl_pytorch/noisy_slos.py:156]: It does what it is expected:

import torch

input_state_tensor = torch.tensor([2, 0, 0])

positions = torch.arange(len(input_state_tensor), dtype=torch.long)
photon_positions = torch.repeat_interleave(positions, input_state_tensor)
print(photon_positions)

outputs tensor([0,0)

Per Perceval, for input [2,0,0], g2 can fire [2,0,0], [3,0,0] or [4,0,0]

File: merlin/core/sectored_distribution.py (hunk 15 above)
Comment (LF-Vigneux) [reply merlin/core/sectored_distribution.py]: Changed in next commit

Comment (LF-Vigneux) [reply merlin/core/sectored_distribution.py]: Changed in next commit

File: merlin/pcvl_pytorch/noisy_slos.py (hunk 13 above)
Comment (LF-Vigneux) [reply merlin/pcvl_pytorch/noisy_slos.py:216]: Changed in next commit

File: merlin/core/sectored_distribution.py [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+class SectorResult:
+    """One photon-number sector of a probability output. If keys are not given, the Combinatics.enumerate_states will be used."""
+
+    tensor: torch.Tensor
+    n_modes: int
+    n_photons: int
+    computation_space: ComputationSpace = ComputationSpace.FOCK
+    keys: tuple[tuple[int, ...], ...] = ()
+
+    def __post_init__(self) -> None:
```
Comment (ben9871) [inline merlin/core/sectored_distribution.py:21]: `SectorResult.__post_init__()` always regenerates fixed-photon Fock keys, even when the caller supplies explicit `keys`. That makes the documented/public `keys` argument unusable and prevents transformed sectors from carrying photon-loss or detector output keys.

File: merlin/algorithms/layer.py [hunk 17]
```diff
@@ -1254,22 +1451,54 @@ def _normalize_output_key(
...
+    def _apply_photon_loss_transform(
+        self, distribution: torch.Tensor | SectoredDistribution
+    ) -> torch.Tensor | SectoredDistribution:
         if self._photon_loss_transform is None:
             raise RuntimeError(
                 "Photon loss transform must be initialised before applying photon loss."
             )
         if self._photon_loss_is_identity:
             return distribution
+        if isinstance(distribution, SectoredDistribution):
```
Comment (ben9871) [inline merlin/algorithms/layer.py:1463]: The per-sector photon-loss transform changes each sector tensor, but the corresponding sector keys are not updated. The returned sector can have `tensor.shape[-1] != len(keys)`, so consumers cannot interpret the probabilities.

File: merlin/algorithms/layer.py [hunk 18]
```diff
@@ -1163,10 +1254,22 @@ def to(self, *args, **kwargs):
 
             # Photon loss Module
             if self._photon_loss_transform is not None:
-                self._photon_loss_transform = self._photon_loss_transform.to(device)
+                if isinstance(self._photon_loss_transform, Sequence):
+                    for i in range(self._photon_loss_transform):
```
Comment (ben9871) [inline merlin/algorithms/layer.py]: This loops with `range(self._photon_loss_transform)` and `range(self._detector_transform)`, but those values are lists in the g2 path. `QuantumLayer.to()` crashes before moving the transforms. Please iterate with `enumerate(self._photon_loss_transform)` / `enumerate(self._detector_transform)` or over `range(len(...))`.

**Review by ben9871 — CHANGES_REQUESTED** [review]

File: tests/pcvl_pytorch/test_noisy_g2_slos.py [hunk 19]
```diff
@@ -1016,3 +1022,81 @@ def test_sectored_total_probability_stays_in_autograd() -> None:
...
+    Correct result: when a caller passes ``keys=...`` to ``SectorResult``, those
+    keys should remain attached to the sector. Auto-generating Fock keys is useful
+    only when ``keys`` is omitted.
+
+    Why this matters: after photon loss or detector transforms, a sector tensor
+    can live in a different output basis from the fixed ``n_photons`` Fock basis.
+    For example, loss from a one-photon sector can include the vacuum key
+    ``(0, 0, 0)``. The container needs to preserve those transform-produced keys.
+
+    Current result: ``SectorResult.__post_init__`` always regenerates the fixed
```
Comment (ben9871) [inline tests/pcvl_pytorch/test_noisy_g2_slos.py]: remove 'current result' from these tests. They are fine now so this is misleading

**Review by ben9871 — CHANGES_REQUESTED** [review]
misleading docstrings in some tests

**Review by ben9871 — APPROVED** [review]
changes made as required. approved
